dvc/repo/tree.py

```diff
@@ -251,14 +251,10 @@ def open(self, path, mode="r", encoding="utf-8", **kwargs):
         if "b" in mode:
             encoding = None
 
-        if self.dvctree and self.dvctree.exists(path):
-            try:
-                return self.dvctree.open(
-                    path, mode=mode, encoding=encoding, **kwargs
-                )
-            except FileNotFoundError:
-                if self.isdvc(path):
-                    raise
+        if self.dvctree and not self.repo.tree.exists(path):
+            return self.dvctree.open(
+                path, mode=mode, encoding=encoding, **kwargs
+            )
```

Comment:
Was also thinking about moving this logic to `DvcTree.open`, but not quite sure. What do you think @pmrowla ?

Reply:
I think moving this logic into `DvcTree.open` would work better. The `tests/func/test_update.py` test failure is because we need to always prefer `dvctree.open` over `repo.tree.open` for DVC outs.

For erepo import/get, `remote.save()` expects `dvctree.open` to fetch DVC outs into local cache as needed when the tree was initialized with `fetch=True`, so if we default to preferring `repo.tree.open` it will not do the fetch (it just calls builtin `open()` via CleanTree/WorkingTree)

If we wanted to keep this logic in `RepoTree` and not `DvcTree`, we would need something like
```py
        if self.dvctree and self.dvctree.exists(path):
            try:
                return self.dvctree.open(
                    path, mode=mode, encoding=encoding, **kwargs
                )
            except FileNotFoundError:
                pass
        return self.repo.tree.open(path)
```

But I think it would make more sense to handle this in `DvcTree.open` with something like
1. open cached version of dvc out when possible (same as existing `DvcTree.open` behavior)
2. try opening (dirty) working copy version of the dvc out if it exists as fallback
3. raise FileNotFoundError

Reply:
For the record: discussed privately and looks like we do agree with the repo tree tests down below, that state that `dirty` version should actually be prefered over the cached one. We got a bit derailed by the failing test, but it might just be a bad test, I'll take a closer look.